src/function/gds/strongly_connected_components.cpp

```diff
@@ -0,0 +1,349 @@
+using namespace kuzu::common;
+using namespace kuzu::processor;
+using namespace kuzu::storage;
+using namespace kuzu::graph;
+
+namespace kuzu {
+namespace function {
+
+static constexpr offset_t SCC_UNSET = std::numeric_limits<offset_t>::max();
+
+struct SCCComputationState {
+    bool allSccIdsSet;
+    ObjectArray<offset_t> sccIDs;
+    AtomicObjectArray<offset_t> fwdWccIDs;
+    AtomicObjectArray<offset_t> bwdWccIDs;
+
+    SCCComputationState(const offset_t numNodes, MemoryManager* mm)
+        : allSccIdsSet{false}, sccIDs{ObjectArray<offset_t>(numNodes, mm)},
+          fwdWccIDs{AtomicObjectArray<offset_t>(numNodes, mm)},
+          bwdWccIDs{AtomicObjectArray<offset_t>(numNodes, mm)} {}
+
+    bool update(offset_t boundOffset, offset_t nbrOffset, bool isFwd) {
+        auto& wccIds = isFwd ? fwdWccIDs : bwdWccIDs;
+        auto boundValue = wccIds.get(boundOffset);
+        auto nbrValue = wccIds.get(nbrOffset);
+        while (nbrValue < boundValue) {
+            if (wccIds.compare_exchange_strong(nbrOffset, nbrValue, boundValue)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool isSccIdSet(offset_t nodeId) { return sccIDs.get(nodeId) != SCC_UNSET; }
+};
+
+class SetInitialSccIds : public VertexCompute {
+public:
+    SetInitialSccIds(SCCComputationState& computationState) : computationState{computationState} {}
+
+    bool beginOnTable(table_id_t) override { return true; }
+
+    void vertexCompute(offset_t startOffset, offset_t endOffset, table_id_t) override {
+        for (auto i = startOffset; i < endOffset; ++i) {
+            computationState.sccIDs.set(i, SCC_UNSET);
+        }
+    }
+
+    std::unique_ptr<VertexCompute> copy() override {
+        return std::make_unique<SetInitialSccIds>(computationState);
+    }
+
+private:
+    SCCComputationState& computationState;
+};
+
+class SetNewSccIds : public VertexCompute {
+public:
+    SetNewSccIds(SCCComputationState& computationState) : computationState{computationState} {}
+
+    bool beginOnTable(table_id_t) override { return true; }
+
+    void vertexCompute(offset_t startOffset, offset_t endOffset, table_id_t) override {
+        for (auto i = startOffset; i < endOffset; ++i) {
+            if (!computationState.isSccIdSet(i)) {
+                auto fwdColor = computationState.fwdWccIDs.get(i);
+                auto bwdColor = computationState.bwdWccIDs.get(i);
+                if (fwdColor == bwdColor) {
+                    computationState.sccIDs.set(i, fwdColor);
+                } else {
+                    // If any of the threads find an unset scc id, they will set the flag to false.
+                    // `allSccIdsSet` is never read and written concurrently, so it does
+                    // not need to be atomic.
+                    computationState.allSccIdsSet = false;
+                }
+            }
+        }
+    }
+
+    void reset() { computationState.allSccIdsSet = true; }
```

Comment:
For this reset function, you should probably use memory_order_seq_cst, so you guarantee it's visible to every thread when the iterations start.